src/gsy_e/gsy_e_core/simulation.py

```diff
@@ -523,9 +543,55 @@ def _get_current_market_time_slot(self, slot_number: int) -> DateTime:
         return (self.area.config.start_date + (slot_number * self.area.config.slot_length)
                 if GlobalConfig.IS_CANARY_NETWORK else self.area.now)
 
+    def _execute_coefficient_simulation(
+            self, slot_resume: int, tick_resume: int, console: NonBlockingConsole = None) -> None:
+
+        slot_count, slot_resume, tick_resume = (
+            self._time.calculate_total_initial_ticks_slots(
+                self._setup.config, slot_resume, tick_resume, self.area))
+
+        for slot_no in range(slot_resume, slot_count):
+            self._handle_paused(console)
+
+            self.progress_info.update(slot_no, slot_count, self._time, self._setup.config)
+
+            self.area.cycle_coefficients_trading(self.progress_info.current_slot_time)
+
+            global_objects.profiles_handler.update_time_and_buffer_profiles(
+                self._get_current_market_time_slot(slot_no))
+
+            self._results.update_and_send_results(
+                self.current_state, self.progress_info, self.area, self._status.status)
+            self._external_events.update(self.area)
+
+            gc.collect()
```

Comment:
Could you explain what objects are in memory and unreachable at this point? I've never used `gc.collect()` explicitly before.

Reply:
This is admittedly not as useful in CPython, but more useful in Pypy due to its aggressive caching behavior. When we encountered issues with severe memory leaks in the gsy-e in the past, we observed that most of the leaks appeared in Pypy and not in CPython. Memory profiling revealed that Pypy in some cases did not delete even objects that were reference counted to zero, in order to cache them and be able to fetch them from its cache if needed later. At some point in time it indeed deallocated them, but not until it had exhausted the memory of the system. Therefore the simulation process was allocating the full amount of memory for prolonged periods of time, which affected the runtime performance of the process. In order to force Pypy to deallocate the memory, we were obliged to force the garbage collector to run on regular intervals, and the ideal place for this operation was on every market cycle. 

We observed also some minor improvements on the CPython version but I suspected that this was due to circular reference detection and not due to the actual garbage collection process, therefore I cannot tell with certainty. Also keep in mind that we run these tests with CPython and Pypy 3.5 version, therefore things might have changed meanwhile. Despite that, I would feel much safer if we left the GC call in place, because worst case scenario we will just trigger an extra collection cycle, which should have minimal effect in the runtime of the application. 

To the feature to be implemented though, I suspect that the memory management will be far far easier, because I expect to have much less cyclical references (due to lack of MarketAgent objects which held references to the Area objects). It remains to be seen.